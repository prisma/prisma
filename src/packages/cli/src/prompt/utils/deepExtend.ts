// Taken from https://github.com/unclechu/node-deep-extend/blob/master/lib/deep-extend.js
// es2017-ified by Tim Suchanek, now it's about 2.5 times faster
/*!
 * @description Recursive object extending
 * @author Viacheslav Lotsmanov <lotsmanov89@gmail.com>
 * @license MIT
 *
 * The MIT License (MIT)
 *
 * Copyright (c) 2013-2018 Viacheslav Lotsmanov
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

function isSpecificValue(val): boolean {
  return val instanceof Buffer || val instanceof Date || val instanceof RegExp
    ? true
    : false
}

function cloneSpecificValue(val): RegExp | Buffer | Date {
  if (val instanceof Buffer) {
    const x = Buffer.alloc ? Buffer.alloc(val.length) : new Buffer(val.length)
    val.copy(x)
    return x
  } else if (val instanceof Date) {
    return new Date(val.getTime())
  } else if (val instanceof RegExp) {
    return new RegExp(val)
  } else {
    throw new Error('Unexpected situation')
  }
}

/**
 * Recursive cloning array.
 */
function deepCloneArray(arr): any {
  const clone: any = []
  arr.forEach(function (item, index) {
    if (typeof item === 'object' && item !== null) {
      if (Array.isArray(item)) {
        clone[index] = deepCloneArray(item)
      } else if (isSpecificValue(item)) {
        clone[index] = cloneSpecificValue(item)
      } else {
        clone[index] = deepExtend({}, item)
      }
    } else {
      clone[index] = item
    }
  })
  return clone
}

function safeGetProperty(object, property): any {
  return property === '__proto__' ? undefined : object[property]
}

/**
 * Extening object that entered in first argument.
 *
 * Returns extended object or false if have no target object or incorrect type.
 *
 * If you wish to clone source object (without modify it), just use empty new
 * object as first argument, like this:
 *   deepExtend({}, yourObj_1, [yourObj_N]);
 */
export const deepExtend = function (target, ...args): any {
  if (!target || typeof target !== 'object') {
    return false
  }

  if (args.length === 0) {
    return target
  }

  let val, src

  for (const obj of args) {
    // skip argument if isn't an object, is null, or is an array
    if (typeof obj !== 'object' || obj === null || Array.isArray(obj)) {
      continue
    }

    for (const key of Object.keys(obj)) {
      src = safeGetProperty(target, key) // source value
      val = safeGetProperty(obj, key) // new value

      // recursion prevention
      if (val === target) {
        continue

        /**
         * if new value isn't object then just overwrite by new value
         * instead of extending.
         */
      } else if (typeof val !== 'object' || val === null) {
        target[key] = val
        continue

        // just clone arrays (and recursive clone objects inside)
      } else if (Array.isArray(val)) {
        target[key] = deepCloneArray(val)
        continue

        // custom cloning and overwrite for specific objects
      } else if (isSpecificValue(val)) {
        target[key] = cloneSpecificValue(val)
        continue

        // overwrite by new value if source isn't object or array
      } else if (
        typeof src !== 'object' ||
        src === null ||
        Array.isArray(src)
      ) {
        target[key] = deepExtend({}, val)
        continue

        // source value and new value is objects both, extending...
      } else {
        target[key] = deepExtend(src, val)
        continue
      }
    }
  }

  return target
}
