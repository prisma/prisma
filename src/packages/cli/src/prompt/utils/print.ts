import chalk from 'chalk'
import { ConnectorType } from '@prisma/generator-helper'

export function printError(text): string {
  return chalk.bold.bgRed(' ERROR ') + ' ' + chalk.red(text)
}

export function printFix(text): string {
  return chalk.bgWhite.black.bold(' FIX ') + ' ' + text
}

const beautifyMap = {
  javascript: 'JavaScript',
  typescript: 'TypeScript',
}

export function beautifyLanguage(language: string): string {
  return beautifyMap[language] || language
}

export function prettyDb(
  type: ConnectorType,
): 'mongo' | 'sqlite' | 'MySQL' | 'PostgreSQL' | string {
  switch (type) {
    case 'mysql': {
      return 'MySQL'
    }
    case 'postgresql': {
      return 'PostgreSQL'
    }
  }
  return type
}
