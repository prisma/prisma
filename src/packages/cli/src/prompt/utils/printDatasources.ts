// This is copied from prisma-client-js/runtime/utils. It needs to be moved into a separate package
import indent from 'indent-string'

export type ConnectorType = 'mysql' | 'mongo' | 'sqlite' | 'postgresql' | 'sqlserver' | 'jdbc:sqlserver'

export interface GeneratorConfig {
  name: string
  output: string | null
  provider: string
  config: Record<string, string>
}

export type Datasource =
  | string
  | {
    url: string
    [key: string]: any | undefined
  }

export interface InternalDatasource {
  name: string
  provider: ConnectorType[]
  url: string
  config: any
}

// We could do import { EnvValue } from '../../isdlToDatamodel2'
// but we don't want to pull that into the runtime build
export interface EnvValue {
  fromEnvVar: null | string
  value: string
}

export function printDatasources(
  internalDatasources: InternalDatasource[],
): string {
  return internalDatasources
    .map((d) => String(new InternalDataSourceClass(d)))
    .join('\n\n')
}

const tab = 2

class InternalDataSourceClass {
  constructor(private readonly dataSource: InternalDatasource) { }

  public toString(): string {
    const { dataSource } = this
    const obj = {
      provider: dataSource.provider,
      url: dataSource.url,
    }
    if (dataSource.config && typeof dataSource.config === 'object') {
      Object.assign(obj, dataSource.config)
    }
    return `datasource ${dataSource.name} {
${indent(printDatamodelObject(obj), tab)}
}`
  }
}

export function printDatamodelObject(obj: any): string {
  const maxLength = Object.keys(obj).reduce(
    (max, curr) => Math.max(max, curr.length),
    0,
  )
  return Object.entries(obj)
    .map(
      ([key, value]: [string, any]) =>
        `${key.padEnd(maxLength)} = ${typeof value === 'object' && value && value.value
          ? JSON.stringify(value.value)
          : JSON.stringify(value)
        }`,
    )
    .join('\n')
}
