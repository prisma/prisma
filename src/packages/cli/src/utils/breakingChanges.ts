import { link } from '@prisma/sdk'
import chalk from 'chalk'

export const breakingChangesMessage = `${chalk.yellow.bold(
  'warn',
)} Prisma 2.12.0 has breaking changes.
You can update your code with ${chalk.bold(
  'npx @prisma/codemods update-2.12 ./',
)}
Read more at ${link('https://pris.ly/2.12')}`
