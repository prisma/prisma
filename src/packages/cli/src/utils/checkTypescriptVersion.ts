import { logger, semverLt } from '@prisma/sdk'
import chalk from 'chalk'
import fs from 'fs'
import path from 'path'
import resolvePkg from 'resolve-pkg'
/**
 * Warn, if typescript is below `4.1.0` and is install locally
 * Because Template Literal Types are required for generating Prisma Client types.
 */
export function checkTypeScriptVersion() {
  const minVersion = '4.1.0'
  try {
    const typescriptPath = resolvePkg('typescript', { cwd: process.cwd() })
    const typescriptPkg =
      typescriptPath && path.join(typescriptPath, 'package.json')
    if (typescriptPkg && fs.existsSync(typescriptPkg)) {
      const pjson = require(typescriptPkg)
      const currentVersion = pjson.version
      if (semverLt(currentVersion, minVersion)) {
        logger.warn(
          `Your ${chalk.bold(
            'typescript',
          )} version is ${currentVersion}, which is outdated. Please update it to ${chalk.bold(
            minVersion,
          )} or ${chalk.bold('newer')} in order to use Prisma Client.`,
        )
      }
    }
  } catch (e) {
    // They do not have TS installed, we ignore (example: JS project)
  }
}
