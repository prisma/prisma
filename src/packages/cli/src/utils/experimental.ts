import chalk from 'chalk'

export class ExperimentalFlagError extends Error {
  constructor() {
    super(
      `Please provide the ${chalk.green(
        '--experimental',
      )} flag to use this command.`,
    )
  }
}
