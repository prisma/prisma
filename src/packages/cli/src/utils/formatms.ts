export function formatms(ms: number) {
  if (ms < 1000) {
    return `${ms}ms`
  }

  return (ms / 1000).toFixed(2) + 's'
}
