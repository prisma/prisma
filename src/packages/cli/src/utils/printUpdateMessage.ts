import { drawBox, isCurrentBinInstalledGlobally } from '@prisma/sdk'
import chalk from 'chalk'
import { Check } from 'checkpoint-client'

const isPrismaInstalledGlobally = isCurrentBinInstalledGlobally()

export function printUpdateMessage(checkResult: {
  status: 'ok'
  data: Check.Response
}): void {
  console.error(
    drawBox({
      height: 4,
      width: 59,
      str: `\n${chalk.blue('Update available')} ${
        checkResult.data.previous_version
      } -> ${checkResult.data.current_version}\nRun the following to update
  ${chalk.bold(
    makeInstallCommand(checkResult.data.package, checkResult.data.release_tag),
  )}
  ${chalk.bold(
    makeInstallCommand('@prisma/client', checkResult.data.release_tag, {
      canBeGlobal: false,
      canBeDev: false,
    }),
  )}`,
      horizontalPadding: 2,
    }),
  )
}

function makeInstallCommand(
  packageName: string,
  tag: string,
  options = {
    canBeGlobal: true,
    canBeDev: true,
  },
): string {
  // Examples
  // yarn 'yarn/1.22.4 npm/? node/v12.14.1 darwin x64'
  // npm 'npm/6.14.7 node/v12.14.1 darwin x64'
  const yarnUsed = process.env.npm_config_user_agent?.includes('yarn')

  let command = ''
  if (isPrismaInstalledGlobally === 'yarn' && options.canBeGlobal) {
    command = `yarn global add ${packageName}`
  } else if (isPrismaInstalledGlobally === 'npm' && options.canBeGlobal) {
    command = `npm i -g ${packageName}`
  } else if (yarnUsed && options.canBeDev) {
    command = `yarn add --dev ${packageName}`
  } else if (options.canBeDev) {
    command = `npm i --save-dev ${packageName}`
  } else if (yarnUsed) {
    command = `yarn add ${packageName}`
  } else {
    command = `npm i ${packageName}`
  }
  if (tag && tag !== 'latest') {
    command += `@${tag}`
  }

  return command
}
