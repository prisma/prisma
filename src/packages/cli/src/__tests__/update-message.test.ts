import 'ts-node/register'
import { printUpdateMessage } from '../utils/printUpdateMessage'
import { consoleContext, Context } from './__helpers__/context'

const ctx = Context.new().add(consoleContext()).assemble()

it('should display a update message w/ dev tag', async () => {
  printUpdateMessage({
    status: 'ok',
    // @ts-ignore
    data: {
      previous_version: '2.6.1',
      current_version: '2.7.0',
      package: '@prisma/cli',
      release_tag: 'dev',
    },
  })
  // @ts-ignore
  const message = ctx.mocked['console.error'].mock.calls[0][0]

  // process.stdout.write(message + "\n")
  expect(message).toMatchSnapshot()
})

it('should display a update message w/o tag', async () => {
  printUpdateMessage({
    status: 'ok',
    // @ts-ignore
    data: {
      previous_version: '2.6.1',
      current_version: '2.7.0',
      package: '@prisma/cli',
      release_tag: 'latest',
    },
  })
  // @ts-ignore
  const message = ctx.mocked['console.error'].mock.calls[0][0]

  // process.stdout.write(message + "\n")
  expect(message).toMatchSnapshot()
})
