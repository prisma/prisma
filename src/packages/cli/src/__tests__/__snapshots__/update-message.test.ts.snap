// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display a update message w/ dev tag 1`] = `
┌─────────────────────────────────────────────────────────┐
│  Update available 2.6.1 -> 2.7.0                        │
│  Run the following to update                            │
│    yarn add --dev @prisma/cli@dev                       │
│    yarn add @prisma/client@dev                          │
└─────────────────────────────────────────────────────────┘
`;

exports[`should display a update message w/o tag 1`] = `
┌─────────────────────────────────────────────────────────┐
│  Update available 2.6.1 -> 2.7.0                        │
│  Run the following to update                            │
│    yarn add --dev @prisma/cli                           │
│    yarn add @prisma/client                              │
└─────────────────────────────────────────────────────────┘
`;
